disnake/ui/_types.py

```diff
@@ -0,0 +1,77 @@
+# SPDX-License-Identifier: MIT
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, NoReturn, Optional, Sequence, TypeVar, Union
+
+if TYPE_CHECKING:
+    from typing_extensions import TypeAlias
+
+    from . import (
+        ActionRow,
+        Button,
+        Container,
+        File,
+        MediaGallery,
+        Section,
+        Separator,
+        TextDisplay,
+        TextInput,
+    )
+    from .item import WrappedComponent
+    from .select import ChannelSelect, MentionableSelect, RoleSelect, StringSelect, UserSelect
+    from .view import View
+
+# TODO: consider if there are any useful types to make public (e.g. disnake-compass used MessageUIComponent)
+
+V_co = TypeVar("V_co", bound="Optional[View]", covariant=True)
+
+AnySelect = Union[
+    "ChannelSelect[V_co]",
+    "MentionableSelect[V_co]",
+    "RoleSelect[V_co]",
+    "StringSelect[V_co]",
+    "UserSelect[V_co]",
+]
+
+# valid `ActionRow.components` item types in a message/modal
+ActionRowMessageComponent = Union["Button[Any]", "AnySelect[Any]"]
+ActionRowModalComponent: TypeAlias = "TextInput"
+
+# valid message component types (v1/v2)
+MessageTopLevelComponentV1: TypeAlias = "ActionRow[ActionRowMessageComponent]"
+MessageTopLevelComponentV2 = Union[
+    "Section",
+    "TextDisplay",
+    "MediaGallery",
+    "File",
+    "Separator",
+    "Container",
+]
+MessageTopLevelComponent = Union[MessageTopLevelComponentV1, MessageTopLevelComponentV2]
+
+ActionRowChildT = TypeVar("ActionRowChildT", bound="WrappedComponent")
```

Comment:
I think this is better placed in `action_row.py`

Reply:
This is being used here as well, I'd like to avoid having imports of type aliases and such from other modules into `_types.py` if possible:
https://github.com/DisnakeDev/disnake/blob/58b60452316c724b5eda648cbab3707654811565/disnake/ui/_types.py#L66-L74